sky/services/raw_keyboard/src/org/chromium/mojo/raw_keyboard/RawKeyboardServiceState.java

```diff
@@ -0,0 +1,76 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package org.chromium.mojo.raw_keyboard;
+
+import android.util.Log;
+import android.view.KeyEvent;
+import android.view.View;
+
+import java.util.ArrayList;
+
+import org.chromium.mojo.system.MojoException;
+import org.chromium.mojom.raw_keyboard.RawKeyboardListener;
+import org.chromium.mojom.raw_keyboard.RawKeyboardService;
+import org.chromium.mojom.sky.InputEvent;
+import org.chromium.mojom.sky.EventType;
+import org.chromium.mojom.sky.KeyData;
+
+/**
+ * Per-View raw keyboard state.
+ */
+public class RawKeyboardServiceState implements View.OnKeyListener {
+    private static final String TAG = "RawKeyboardServiceState";
+
+    private View mView;
+    private ArrayList<RawKeyboardListener> mListeners = new ArrayList<RawKeyboardListener>();
+
+    public RawKeyboardServiceState(View view) {
+        mView = view;
+        mView.setOnKeyListener(this);
+    }
+
+    public void addListener(RawKeyboardListener listener) {
+        mListeners.add(listener);
+    }
+
+    public void removeListener(RawKeyboardListener listener) {
+        mListeners.remove(listener);
+        listener.close();
+    }
+
+    public void close() {
+        mView.setOnKeyListener(null);
+        for(RawKeyboardListener listener : mListeners)
+            removeListener(listener);
+    }
+
+    @Override
+    public boolean onKey(View v, int keyCode, KeyEvent nativeEvent) {
+         if (mListeners.isEmpty())
+             return false;
+         InputEvent event = new InputEvent();
+         switch (nativeEvent.getAction()) {
+           case KeyEvent.ACTION_DOWN:
+               event.action = EventType.KEY_PRESSED;
+               break;
+           case KeyEvent.ACTION_UP:
+               event.action = EventType.KEY_RELEASED;
+               break;
+           default:
+               Log.w(TAG, "Unknown key event received");
+               return false;
+         }
+         KeyData keyData = new KeyData();
+         keyData.flags = nativeEvent.getFlags();
+         keyData.scanCode = nativeEvent.getScanCode();
+         keyData.metaState = nativeEvent.getMetaState();
+         keyData.keyCode = keyCode;
+         event.keyData = keyData;
+         Log.v(TAG, "Key event recorded");
+         for (RawKeyboardListener listener : mListeners)
+             listener.onKey(event);
+         return true;
```

Comment:
Default behaviors aren't an immediate concern as we are just trying to put together a PoC of hardware input at this point, but I agree that the API shouldn't be considered ready until we switch to using the NDK APIs. I'll file an issue for that once this is landed.